plaidml/bridge/openvino/ops/roi_pooling.cpp

```diff
@@ -14,70 +14,109 @@ using namespace InferenceEngine;  // NOLINT[build/namespaces]
-
-  auto roi_width = y_2 - y_1;
-  auto roi_height = x_2 - x_1;
-
-  auto roi_h_scale = roi_height / pooled_h;
-  auto roi_w_scale = roi_width / pooled_w;
-
-  auto h_tensor = edsl::cast(edsl::index({edsl::TensorDim(pooled_h)}, 0), DType::FLOAT32) * roi_h_scale + x_1;
-  auto w_tensor = edsl::cast(edsl::index({edsl::TensorDim(pooled_w)}, 0), DType::FLOAT32) * roi_w_scale + y_1;
-
-  auto gather_w = edsl::gather(I, w_tensor).axis(3).interpolationMode(edsl::InterpolationMode::LINEAR);
-  auto gather_h = edsl::gather(gather_w, h_tensor).axis(2).interpolationMode(edsl::InterpolationMode::LINEAR);
-
-  return gather_h;
+edsl::Tensor bilinear_pooling(edsl::Tensor I, const std::vector<float>& coord, int64_t pooled_h, int64_t pooled_w,
+                              int height, int width) {
+  auto roi_w_start = coord[0];
+  auto roi_h_start = coord[1];
+  auto roi_w_end = coord[2];
+  auto roi_h_end = coord[3];
+
+  float roi_width = (roi_w_end - roi_w_start) * (width - 1);
+  float roi_height = (roi_h_end - roi_h_start) * (height - 1);
+
+  float roi_h_scale = roi_height / (pooled_h - 1);
+  float roi_w_scale = roi_width / (pooled_w - 1);
+
+  // get center point of every ROI bin.
+  auto in_y = edsl::cast(edsl::index({edsl::TensorDim(pooled_h)}, 0), DType::FLOAT32) * roi_h_scale +
+              roi_h_start * (height - 1);
+  auto in_x =
+      edsl::cast(edsl::index({edsl::TensorDim(pooled_w)}, 0), DType::FLOAT32) * roi_w_scale + roi_w_start * (width - 1);
+
+  // get it's interpolation point.
+  auto top_y_index = edsl::floor(in_y);
+  auto bottom_y_index = edsl::ceil(in_y);
+  auto left_x_index = edsl::floor(in_x);
+  auto right_x_index = edsl::ceil(in_x);
+
+  // check it border.
+  auto h_border = edsl::cast(edsl::Tensor(height - 1), DType::FLOAT32);
+  auto w_border = edsl::cast(edsl::Tensor(width - 1), DType::FLOAT32);
+  bottom_y_index = edsl::select(bottom_y_index < h_border, bottom_y_index, h_border);
+  right_x_index = edsl::select(right_x_index < w_border, right_x_index, w_border);
+
+  // gather the interpolation point from feature map.
+  auto get_pieces = [](edsl::Tensor temp, edsl::Tensor y, edsl::Tensor x) {
+    auto gather_x = edsl::gather(temp, x).axis(3);
+    auto gather_y = edsl::gather(gather_x, y).axis(2);
+    return gather_y;
+  };
+  edsl::Tensor top_left = get_pieces(I, top_y_index, left_x_index);
+  edsl::Tensor top_right = get_pieces(I, top_y_index, right_x_index);
+  edsl::Tensor bottom_left = get_pieces(I, bottom_y_index, left_x_index);
+  edsl::Tensor bottom_right = get_pieces(I, bottom_y_index, right_x_index);
+
+  std::vector<edsl::TensorDim> dims(I.rank());
+  I.bind_dims(dims);
+  dims[2] = edsl::TensorDim(pooled_h);
+  dims[3] = edsl::TensorDim(pooled_w);
+  auto top = top_left + (top_right - top_left) * (in_x - left_x_index);
+  auto bottom = bottom_left + (bottom_right - bottom_left) * (in_x - left_x_index);
+  // output = top + (bottom - top) * (in_y - top_y_index), we have to reshape mul openrands.
```

Comment:
Probably it is better to add a new line before this comment.

Reply:
Typo: operands

Reply:
done